### PR TITLE
Expose the set of tags to reverse and reverse complement as parameters.

### PR DESCRIPTION
**Pull request by tfenne**
### Description

Expose the set of tags to reverse and reverse complement as parameters, so that reads can be reverse complemented and all tags that are one-value-per-base in the read can be kept in sync.

### Checklist

- [x] Code compiles correctly
- [x] New tests covering changes and new functionality
- [x] All tests passing
- [x] Extended the README / documentation, if necessary
- [x] Is not backward compatible (breaks binary or source compatibility)

### CODE REVIEW

File: src/main/java/htsjdk/samtools/SAMRecordUtil.java [hunk 1]
```diff
@@ -26,56 +26,96 @@
...
+    private static List<String> TAGS_TO_REVERSE            = Arrays.asList(SAMTag.OQ.name(), SAMTag.U2.name());
 
     /**
      * Reverse-complement all known sequence and base quality attributes of the SAMRecord.
      */
     public static void reverseComplement(final SAMRecord rec) {
+        reverseComplement(rec, TAGS_TO_REVERSE_COMPLEMENT, TAGS_TO_REVERSE);
+    }
+
+    public static void reverseComplement(final SAMRecord rec, Collection<String> tagsToRevcomp, Collection<String> tagsToReverse) {
```
Comment (eitanbanks) [inline src/main/java/htsjdk/samtools/SAMRecordUtil.java]: Need 'final's here

File: src/main/java/htsjdk/samtools/SAMRecordUtil.java [hunk 2]
```diff
@@ -26,56 +26,96 @@
...
+
+        // Deal with tags that needed to just be reversed
+        if (tagsToReverse != null) {
+            for (final String tag : tagsToReverse) {
+                Object value = rec.getAttribute(tag);
+                if (value != null) {
+                    if (value instanceof String) {
+                        value = StringUtil.reverseString((String) value);
+                    }
+                    else if (value.getClass().isArray()) {
```
Comment (eitanbanks) [inline src/main/java/htsjdk/samtools/SAMRecordUtil.java]: Just out of curiosity, why not maintain the previous behavior (converting to String and then reversing that)?  Code was a lot cleaner.  Is it b/c of bad performance?
Maybe pull all this out into a reverseArray() method that takes in an Object for code cleanliness?

**Comment by eitanbanks** [comment]
Comments made.  @nh13 should feel free to comment himself (as the assigned reviewer) or just assign to me.

File: src/main/java/htsjdk/samtools/SAMRecordUtil.java (hunk 2 above)
Comment (nh13) [inline src/main/java/htsjdk/samtools/SAMRecordUtil.java]: It actually only supported `String` and not the array types.  The increased complexity is to handle the various array types.  I am not sure moving the 5 lines of code into a seperate `reverseArray` method is really cleaner.

**Comment by nh13** [comment]
@tfenne 👍 me but please respond to @eitanbanks comments.

File: src/main/java/htsjdk/samtools/SAMRecordUtil.java [hunk 3]
```diff
@@ -26,56 +26,96 @@
...
-    /** List of String tags that must be reversed if present when a SAMRecord is reverseComplemented */
-    private static final short[] STRING_TAGS_TO_REVERSE = {
-            SAMTagUtil.getSingleton().U2,
-            SAMTagUtil.getSingleton().OQ
-    };
+    private static List<String> TAGS_TO_REVERSE_COMPLEMENT = Arrays.asList(SAMTag.E2.name(), SAMTag.SQ.name());
+    private static List<String> TAGS_TO_REVERSE            = Arrays.asList(SAMTag.OQ.name(), SAMTag.U2.name());
 
     /**
      * Reverse-complement all known sequence and base quality attributes of the SAMRecord.
```
Comment (yfarjoun) [inline src/main/java/htsjdk/samtools/SAMRecordUtil.java]: update javadoc

File: src/main/java/htsjdk/samtools/SAMRecordUtil.java [hunk 4]
```diff
@@ -26,56 +26,96 @@
...
+                    if (value instanceof String) {
+                        value = StringUtil.reverseString((String) value);
+                    }
+                    else if (value.getClass().isArray()) {
+                        if (value instanceof byte[]) reverseArray((byte[]) value);
+                        else if (value instanceof short[]) reverseArray((short[]) value);
+                        else if (value instanceof int[]) reverseArray((int[]) value);
+                        else if (value instanceof float[]) reverseArray((float[]) value);
+                        else throw new UnsupportedOperationException("Reversing array attribute of type " + value.getClass().getComponentType() + " not supported.");
+                    }
```
Comment (yfarjoun) [inline src/main/java/htsjdk/samtools/SAMRecordUtil.java]: shouldn't there be another else clause here?

File: src/main/java/htsjdk/samtools/SAMRecordUtil.java (hunk 3 above)
Comment (tfenne) [inline src/main/java/htsjdk/samtools/SAMRecordUtil.java]: Done.

File: src/main/java/htsjdk/samtools/SAMRecordUtil.java (hunk 1 above)
Comment (tfenne) [inline src/main/java/htsjdk/samtools/SAMRecordUtil.java]: Done.

File: src/main/java/htsjdk/samtools/SAMRecordUtil.java (hunk 4 above)
Comment (tfenne) [inline src/main/java/htsjdk/samtools/SAMRecordUtil.java]: Yes, done.

**Comment by tfenne** [comment]
Looks like just the `remote` tests are failing (again), and nothing to do with this PR.  I've re-kicked the build, but if they fail again I'll merge anyway.

**Pull request merged**